src/absil/illib.fs

```diff
@@ -448,6 +448,46 @@ module Dictionary =
 module Lazy = 
     let force (x: Lazy<'T>) = x.Force()
 
+//----------------------------------------------------------------------------
+// Singe threaded execution and mutual exclusion
+
+/// Represents a permission active at this point in execution
+type ExecutionToken = interface end
+
+/// Represents a token that indicates execution on the compilation thread, i.e. 
+///   - we have full access to the (partially mutable) TAST and TcImports data structures
+///   - compiler execution may result in type provider invocations when resolving types and members
+///   - we can access various caches in the SourceCodeServices
+///
+/// Like other execution tokens this should be passed via argument passing and not captured/stored beyond
+/// the lifetime of stack-based calls. This is not checked, it is a discipline withinn the compiler code. 
+type CompilationThreadToken() = interface ExecutionToken
+
+/// Represnts a place where we are stating that execution on the compilation thread is required.  The
+/// reason why will be documented in a comment in the code at the callsite.
+let RequireCompilationThread (_ctok: CompilationThreadToken) = ()
+
+/// Represnts a place in the compiler codebase where we are passed a CompilationThreadToken unnecessarily.
+/// This reprents code that may potentially not need to be executed on the compilation thread.
+let DoesNotSpecificallyRequireCompilerThreadAndCouldLikelyBeConcurrent (_ctok: CompilationThreadToken) = ()
```

Comment:
The name of this isn't particularly succinct or clear. What does 'specifically require' mean? Surely it either does require it, or it doesn't - it's binary. Unless there's some grey area that I'm missing. The same applies for 'CouldLikelyBeConcurrent' - is it thread-safe or not?